interfaces/builtin/appstream_metadata.go

```diff
@@ -0,0 +1,120 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/mount"
+	"github.com/snapcore/snapd/osutil"
+)
+
+const appstreamMetadataSummary = `allows access to AppStream metadata`
+
+const appstreamMetadataBaseDeclarationSlots = `
+  appstream-metadata:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+// Paths for upstream and collection metadata are defined in the
+// AppStream specification:
+//   https://www.freedesktop.org/software/appstream/docs/
+const appstreamMetadataConnectedPlugAppArmor = `
+# Description: Allow access to AppStream metadata from the host system
+
+# Allow access to AppStream upstream metadata files
+/usr/share/metainfo/** r,
+/usr/share/appdata/** r,
+
+# Allow access to AppStream collection metadata
+/usr/share/app-info/** r,
+/var/cache/app-info/** r,
+/var/lib/app-info/** r,
+
+# Apt symlinks the DEP-11 metadata to files in /var/lib/apt/lists
+/var/lib/apt/lists/*.yml.gz r,
+`
+
+var appstreamMetadataDirs = []string{
+	"/usr/share/metainfo",
+	"/usr/share/appdata",
+	"/usr/share/app-info",
+	"/var/cache/app-info",
+	"/var/lib/app-info",
+	"/var/lib/apt/lists",
```

Comment:
Should this include support for other package managers or is this just due to some library that does this as a matter of course?

Reply:
As mentioned in source comments, we aren't exposing Apt specific metadata.  The problem is that Apt installs AppStream metadata to `/var/lib/apt/lists` and then symlinks it to `/var/lib/app-info`.  If we don't bind mount `/var/lib/apt/lists` into the sandbox, all we have is a bunch of dangling symlinks.

The AppArmor rules enforce that the only files in `/var/lib/apt/lists` that can actually be read are the AppStream `.yml.gz` files.

Reply:
> As mentioned in source comments, we aren't exposing Apt specific metadata. The problem is that Apt installs AppStream metadata to `/var/lib/apt/lists` and then symlinks it to `/var/lib/app-info`. If we don't bind mount `/var/lib/apt/lists` into the sandbox, all we have is a bunch of dangling symlinks.
> 
> The AppArmor rules enforce that the only files in `/var/lib/apt/lists` that can actually be read are the AppStream `.yml.gz` files.

Sure. My point wasn't about dropping /var/lib/apt/lists, it was about does yum, dnf, pacman, etc do something similar and so should this be extended. "I don't know" is an acceptable answer-- these can come in subsequent PRs as needed.

Reply:
I haven't checked everywhere, but Debian/Ubuntu seem unusual in this respect.

For Fedora, the metadata is shipped as an RPM package installing directly to `/usr/share/app-info/xmls`: https://apps.fedoraproject.org/packages/appstream-data/

I'm not sure what OpenSUSE or Arch do, so I think the follow-up PRs would be a sensible way to move forward if what we have isn't sufficient.

Reply:
Thanks for looking. Based on your response a "wait and see" approach sounds fine to me.